### PR TITLE
Add ROCmPackage to hypre

### PR DESCRIPTION
**Pull request by jrood-nrel**
I've tested this to work on Spock so far.

### CODE REVIEW

**Comment by jrood-nrel** [comment]
I verified my changes to CUDA work on Summit as well. Hoping to get a review on this so I can start building some other things on top of this.

**Comment by jrood-nrel** [comment]
Anybody able to give a review on this?

File: var/spack/repos/builtin/packages/hypre/package.py [hunk 1]
```diff
@@ -93,7 +93,9 @@ class Hypre(AutotoolsPackage, CudaPackage):
     depends_on('superlu-dist', when='+superlu-dist+mpi')
 
     conflicts('+cuda', when='+int64')
-    conflicts('+unified-memory', when='~cuda')
+    conflicts('+rocm', when='+int64')
+    conflicts('+rocm', when='@:2.23.0')
```
Comment (osborn9) [inline var/spack/repos/builtin/packages/hypre/package.py]: HIP support added in v2.21.0
```suggestion
    conflicts('+rocm', when='@:2.20')
```

**Review by osborn9 — CHANGES_REQUESTED** [review]
Modulo the minor fix for the minimum version with HIP support, this PR looks good, thanks!

To digress, hypre is currently assuming `rocrand` and `rocsparse` are installed in the same location as `hip`. As pointed out in #27743, we need to change this within `hypre`, so that  having spack install these dependencies can be supported. 

Anyways, this PR works with the current assumptions of building `hypre` with ROCm.

**Comment by jrood-nrel** [comment]
Thanks for the review. I made the suggested change.

**Review by osborn9 — APPROVED** [review]

**Review by ulrikeyang — APPROVED** [review]

**Comment by osborn9** [comment]
@balay Could you take a look at this? There's still the outstanding issue within hypre (from #27743), which I will prioritize addressing. In the meantime, this will be helpful for users getting workflows on Spock built.

**Review by alalazo — APPROVED** [review]
Already approved by maintainers

**Pull request merged**